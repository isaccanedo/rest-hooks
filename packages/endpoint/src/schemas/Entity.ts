/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { AbstractInstanceType } from '../normal.js';
import type { Schema, NormalizedIndex, UnvisitFunction } from '../interface.js';
import { isImmutable, denormalizeImmutable } from './ImmutableUtils.js';

/**
 * Represents data that should be deduped by specifying a primary key.
 * @see https://resthooks.io/docs/api/Entity
 */
export default abstract class Entity {
  static toJSON() {
    return {
      name: this.name,
      schema: this.schema,
      key: this.key,
    };
  }

  /** Defines nested entities */
  static schema: { [k: string]: Schema } = {};

  /**
   * A unique identifier for each Entity
   *
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   */
  abstract pk(parent?: any, key?: string): string | undefined;

  /** Returns the globally unique identifier for the static Entity */
  static get key(): string {
    /* istanbul ignore next */
    if (
      process.env.NODE_ENV !== 'production' &&
      (this.name === '' || this.name === 'Entity' || this.name === '_temp')
    )
      throw new Error(
        'Entity classes without a name must define `static get key()`',
      );
    return this.name;
  }

  /** Defines indexes to enable lookup by */
  declare static indexes?: readonly string[];

  /** Control how automatic schema validation is handled
   *
   * `undefined`: Defaults - throw error in worst offense
   * 'warn': only ever warn
   * 'silent': Don't bother with processing at all
   *
   * Note: this only applies to non-nested members.
   */
  protected declare static automaticValidation?: 'warn' | 'silent';

  /**
   * A unique identifier for each Entity
   *
   * @param [value] POJO of the entity or subset used
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   */
  static pk<T extends typeof Entity>(
    this: T,
    value: Partial<AbstractInstanceType<T>>,
    parent?: any,
    key?: string,
  ): string | undefined {
    return this.prototype.pk.call(value, parent, key);
  }

  /** Return true to merge incoming data; false keeps existing entity */
  static useIncoming(
    existingMeta: { date: number; fetchedAt: number },
    incomingMeta: { date: number; fetchedAt: number },
    existing: any,
    incoming: any,
  ) {
    return existingMeta.fetchedAt <= incomingMeta.fetchedAt;
  }

  /** Creates new instance copying over defined values of arguments */
  static merge(existing: any, incoming: any) {
    return {
      ...existing,
      ...incoming,
    };
  }

  /** Factory method to convert from Plain JS Objects.
   *
   * @param [props] Plain Object of properties to assign.
   * @param [parent] When normalizing, the object which included the record
   * @param [key] When normalizing, the key where this record was found
   */
  static fromJS<T extends typeof Entity>(
    this: T,
    // TODO: this should only accept members that are not functions
    props: Partial<AbstractInstanceType<T>> = {},
  ): AbstractInstanceType<T> {
    // we type guarded abstract case above, so ok to force typescript to allow constructor call
    const instance = new (this as any)(props) as AbstractInstanceType<T>;
    // we can't rely on constructors and override the defaults provided as property assignments
    // all occur after the constructor
    Object.assign(instance, props);
    return instance;
  }

  /** Do any transformations when first receiving input */
  static process(input: any, parent: any, key: string | undefined): any {
    /* istanbul ignore else */
    if (
      process.env.NODE_ENV !== 'production' &&
      this.automaticValidation !== 'silent'
    ) {
      if (Array.isArray(input)) {
        const errorMessage = `Attempted to initialize ${
          this.name
        } with an array, but named members were expected

This is likely due to a malformed response.
Try inspecting the network response or fetch() return value.
Or use debugging tools: https://resthooks.io/docs/guides/debugging
Learn more about schemas: https://resthooks.io/docs/api/schema
If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

First three members: ${JSON.stringify(input.slice(0, 3), null, 2)}`;
        if (this.automaticValidation !== 'warn') {
          const error = new Error(errorMessage);
          (error as any).status = 400;
          throw error;
        }
        console.warn(errorMessage);
      }
    }

    return { ...input };
  }

  static normalize(
    input: any,
    parent: any,
    key: string | undefined,
    visit: (...args: any) => any,
    addEntity: (...args: any) => any,
    visitedEntities: Record<string, any>,
  ): any {
    const processedEntity = this.process(input, parent, key);
    const id = this.pk(processedEntity, parent, key);
    if (id === undefined || id === '') {
      if (process.env.NODE_ENV !== 'production') {
        const error = new Error(
          `Missing usable primary key when normalizing response.

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.
  Or use debugging tools: https://resthooks.io/docs/guides/debugging
  Learn more about schemas: https://resthooks.io/docs/api/schema

  Entity: ${this.name}
  Value (processed): ${
    processedEntity && JSON.stringify(processedEntity, null, 2)
  }
  `,
        );
        (error as any).status = 400;
        throw error;
      } else {
        // these make the keys get deleted
        return undefined;
      }
    }
    const entityType = this.key;

    if (!(entityType in visitedEntities)) {
      visitedEntities[entityType] = {};
    }
    if (!(id in visitedEntities[entityType])) {
      visitedEntities[entityType][id] = [];
    }
    if (
      visitedEntities[entityType][id].some((entity: any) => entity === input)
    ) {
      return id;
    }
    const errorMessage = this.validate(processedEntity);
    if (errorMessage) {
      const error = new Error(errorMessage);
      (error as any).status = 400;
      throw error;
    }
    visitedEntities[entityType][id].push(input);

    Object.keys(this.schema).forEach(key => {
      if (Object.hasOwn(processedEntity, key)) {
        const schema = this.schema[key];
        processedEntity[key] = visit(
          processedEntity[key],
          processedEntity,
          key,
          schema,
          addEntity,
          visitedEntities,
        );
      }
    });

    addEntity(this, processedEntity, id);
    return id;
  }

  protected static validate(processedEntity: any): string | undefined {
    /* istanbul ignore else */
    if (
      process.env.NODE_ENV !== 'production' &&
      this.automaticValidation !== 'silent'
    ) {
      const keysOfRecord = new Set(Object.keys(this.defaults));
      const keysOfProps = Object.keys(processedEntity);
      const [found, missing, unexpected] = [[], [], []] as [
        string[],
        string[],
        string[],
      ];
      for (const keyOfProps of keysOfProps) {
        if (keysOfRecord.has(keyOfProps)) {
          found.push(keyOfProps);
        } else {
          unexpected.push(keyOfProps);
        }
      }
      for (const keyOfRecord of keysOfRecord) {
        if (!found.includes(keyOfRecord)) {
          missing.push(keyOfRecord);
        }
      }

      // only bother with this if they used *any* defaults
      if (keysOfRecord.size) {
        const tooManyUnexpected =
          // unexpected compared to members in response
          Math.max(keysOfProps.length / 2, 1) <= unexpected.length &&
          // unexpected compared to what we specified
          keysOfRecord.size > Math.max(unexpected.length, 2) &&
          // as we find more and more be more easily assured it is correct
          found.length ** 1.5 / 2 <= unexpected.length;
        const foundNothing = found.length < Math.min(1, keysOfRecord.size / 2);
        // if we find nothing (we expect at least one member for a pk)
        // or we find too many unexpected members
        if (tooManyUnexpected || foundNothing) {
          let extra = '';
          let reason = 'substantially different than expected keys';
          if (foundNothing) {
            extra += `\n    Missing: ${missing}`;
            reason = 'no matching keys found';
          }
          if (tooManyUnexpected) {
            extra += `\n    Unexpected keys: ${unexpected}`;
            reason = 'a large number of unexpected keys found';
          }
          const errorMessage = `Attempted to initialize ${
            this.name
          } with ${reason}

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.
  Or use debugging tools: https://resthooks.io/docs/guides/debugging
  Learn more about schemas: https://resthooks.io/docs/api/schema
  If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

  Expected keys:
    Found: ${found}${extra}
  Value (processed): ${JSON.stringify(processedEntity, null, 2)}`;
          if (
            (found.length >= 4 && tooManyUnexpected) ||
            this.automaticValidation === 'warn'
          ) {
            console.warn(errorMessage);
          } else {
            return errorMessage;
          }
        }
      }
    }
    if (process.env.NODE_ENV !== 'production') {
      for (const key of Object.keys(this.schema)) {
        if (!Object.hasOwn(processedEntity, key)) {
          if (!Object.hasOwn(this.defaults, key)) {
            return `Schema key is missing in Entity

  Be sure all schema members are also part of the entity
  Or use debugging tools: https://resthooks.io/docs/guides/debugging
  Learn more about nesting schemas: https://resthooks.io/docs/guides/nested-response

  Entity keys: ${Object.keys(this.defaults)}
  Schema key(missing): ${key}
  `;
          }
        }
      }
    }
  }

  static infer(
    args: readonly any[],
    indexes: NormalizedIndex,
    recurse: any,
  ): any {
    if (!args[0]) return undefined;
    if (['string', 'number'].includes(typeof args[0])) {
      return `${args[0]}`;
    }
    const id = this.pk(args[0], undefined, '');
    // Was able to infer the entity's primary key from params
    if (id !== undefined && id !== '') return id;
    // now attempt lookup in indexes
    const indexName = indexFromParams(args[0], this.indexes);
    if (indexName && indexes[this.key]) {
      // 'as Record<string, any>': indexName can only be found if params is a string key'd object
      const id =
        indexes[this.key][indexName][
          (args[0] as Record<string, any>)[indexName]
        ];
      return id;
    }
    return undefined;
  }

  static expiresAt(
    meta: { expiresAt: number; date: number; fetchedAt: number },
    input: any,
  ): number {
    return meta.expiresAt;
  }

  static denormalize<T extends typeof Entity>(
    this: T,
    input: any,
    unvisit: UnvisitFunction,
  ): [denormalized: AbstractInstanceType<T>, found: boolean, suspend: boolean] {
    // TODO: remove codecov ignore once denormalize is modified to expect this
    /* istanbul ignore if */
    if (typeof input === 'symbol') {
      return [undefined, true, true] as any;
    }
    if (isImmutable(input)) {
      if (this.validate((input as any).toJS()))
        return [undefined as any, false, true];
      // Need to set this first so that if it is referenced further within the
      // denormalization the reference will already exist.
      unvisit.setLocal?.(input);
      const [denormEntity, found, deleted] = denormalizeImmutable(
        this.schema,
        input,
        unvisit,
      );
      return [this.fromJS(denormEntity.toObject()), found, deleted];
    }
    if (this.validate(input)) {
      return [undefined as any, false, true];
    }
    const entityCopy: any = this.fromJS(input);
    // Need to set this first so that if it is referenced further within the
    // denormalization the reference will already exist.
    unvisit.setLocal?.(entityCopy);

    let deleted = false;

    // note: iteration order must be stable
    Object.keys(this.schema).forEach(key => {
      const schema = this.schema[key];
      const nextInput = Object.hasOwn(input, key)
        ? (input as any)[key]
        : undefined;
      const [value, , deletedItem] = unvisit(nextInput, schema);

      if (deletedItem && !(Object.hasOwn(input, key) && !this.defaults[key])) {
        deleted = true;
      }
      if (Object.hasOwn(input, key) && (input as any)[key] !== value) {
        this.set(entityCopy, key, value);
      }
    });

    return [entityCopy, true, deleted];
  }

  private declare static __defaults: any;
  /** All instance defaults set */
  protected static get defaults() {
    if (!Object.hasOwn(this, '__defaults'))
      this.__defaults = new (this as any)();
    return this.__defaults;
  }

  /** Used by denormalize to set nested members */
  protected static set(entity: any, key: string, value: any) {
    entity[key] = value;
  }
}

/* istanbul ignore else */
if (process.env.NODE_ENV !== 'production') {
  const superFrom = Entity.fromJS;
  // for those not using TypeScript this is a good catch to ensure they are defining
  // the abstract members
  Entity.fromJS = function fromJS<T extends typeof Entity>(
    this: T,
    props: Partial<AbstractInstanceType<T>>,
  ): AbstractInstanceType<T> {
    if ((this as any).prototype.pk === undefined)
      throw new Error('cannot construct on abstract types');
    return superFrom.call(this, props) as any;
  };
}

function indexFromParams<I extends string>(
  params: Readonly<object>,
  indexes?: Readonly<I[]>,
) {
  if (!indexes) return undefined;
  return indexes.find(index => Object.hasOwn(params, index));
}
