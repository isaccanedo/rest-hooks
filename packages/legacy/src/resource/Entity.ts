/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { AbstractInstanceType, Schema, schema } from '@rest-hooks/endpoint';

import { isImmutable, denormalizeImmutable } from './ImmutableUtils.js';
import SimpleRecord from './SimpleRecord.js';

/** Represents data that should be deduped by specifying a primary key. */
export default abstract class Entity extends SimpleRecord {
  static toJSON() {
    return {
      ...super.toJSON(),
      key: this.key,
    };
  }

  /**
   * A unique identifier for each Entity
   *
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   */
  abstract pk(parent?: any, key?: string): string | undefined;

  /** Returns the globally unique identifier for the static Entity */
  static get key(): string {
    /* istanbul ignore next */
    if (
      process.env.NODE_ENV !== 'production' &&
      (this.name === '' || this.name === 'Entity' || this.name === '_temp')
    )
      throw new Error(
        'Entity classes without a name must define `static get key()`',
      );
    return this.name;
  }

  /** Defines indexes to enable lookup by */
  declare static indexes?: readonly string[];

  /** Control how automatic schema validation is handled
   *
   * `undefined`: Defaults - throw error in worst offense
   * 'warn': only ever warn
   * 'silent': Don't bother with processing at all
   *
   * Note: this only applies to non-nested members.
   */
  protected declare static automaticValidation?: 'warn' | 'silent';

  /**
   * A unique identifier for each Entity
   *
   * @param [value] POJO of the entity or subset used
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   */
  static pk<T extends typeof Entity>(
    this: T,
    value: Partial<AbstractInstanceType<T>>,
    parent?: any,
    key?: string,
  ): string | undefined {
    return this.prototype.pk.call(value, parent, key) || key;
  }

  static normalize(
    input: any,
    parent: any,
    key: string | undefined,
    visit: (...args: any) => any,
    addEntity: (...args: any) => any,
    visitedEntities: Record<string, any>,
  ): any {
    // pass over already processed entities
    if (typeof input === 'string') return input;
    // TODO: what's store needs to be a differing type from fromJS
    const processedEntity: any = this.fromJS(input, parent, key);
    /* istanbul ignore else */
    if (
      process.env.NODE_ENV !== 'production' &&
      this.automaticValidation !== 'silent'
    ) {
      const keysOfRecord = new Set(Object.keys(this.defaults));
      const keysOfProps = this.keysDefined(processedEntity);
      const [found, missing, unexpected] = [[], [], []] as [
        string[],
        string[],
        string[],
      ];
      for (const keyOfProps of keysOfProps) {
        if (keysOfRecord.has(keyOfProps)) {
          found.push(keyOfProps);
        } else {
          unexpected.push(keyOfProps);
        }
      }
      for (const keyOfRecord of keysOfRecord) {
        if (!found.includes(keyOfRecord)) {
          missing.push(keyOfRecord);
        }
      }

      // only bother with this if they used *any* defaults
      if (keysOfRecord.size) {
        if (Array.isArray(input) && unexpected.length) {
          const errorMessage = `Attempted to initialize ${
            this.name
          } with an array, but named members were expected

This is likely due to a malformed response.
Try inspecting the network response or fetch() return value.
Or use debugging tools: https://resthooks.io/docs/guides/debugging
Learn more about schemas: https://resthooks.io/docs/api/schema
If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

Missing: ${missing}
First three members: ${JSON.stringify(input.slice(0, 3), null, 2)}`;
          if (this.automaticValidation !== 'warn') {
            const error = new Error(errorMessage);
            (error as any).status = 400;
            throw error;
          }
          console.warn(errorMessage);
        }

        const tooManyUnexpected =
          // unexpected compared to members in response
          Math.max(keysOfProps.length / 2, 1) <= unexpected.length &&
          // unexpected compared to what we specified
          keysOfRecord.size > Math.max(unexpected.length, 2) &&
          // as we find more and more be more easily assured it is correct
          found.length ** 1.5 / 2 <= unexpected.length;
        const foundNothing = found.length < Math.min(1, keysOfRecord.size / 2);
        // if we find nothing (we expect at least one member for a pk)
        // or we find too many unexpected members
        if (tooManyUnexpected || foundNothing) {
          let extra = '';
          let reason = 'substantially different than expected keys';
          if (foundNothing) {
            extra += `\n    Missing: ${missing}`;
            reason = 'no matching keys found';
          }
          if (tooManyUnexpected) {
            extra += `\n    Unexpected keys: ${unexpected}`;
            reason = 'a large number of unexpected keys found';
          }
          const errorMessage = `Attempted to initialize ${
            this.name
          } with ${reason}

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.
  Or use debugging tools: https://resthooks.io/docs/guides/debugging
  Learn more about schemas: https://resthooks.io/docs/api/schema
  If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

  Expected keys:
    Found: ${found}${extra}
  Value: ${JSON.stringify(this.toObjectDefined(processedEntity), null, 2)}`;
          if (
            (found.length >= 4 && tooManyUnexpected) ||
            this.automaticValidation === 'warn'
          ) {
            console.warn(errorMessage);
          } else {
            const error = new Error(errorMessage);
            (error as any).status = 400;
            throw error;
          }
        }
      }
    }
    const id = processedEntity.pk(parent, key);
    if (id === undefined || id === '') {
      if (process.env.NODE_ENV !== 'production') {
        const error = new Error(
          `Missing usable primary key when normalizing response.

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.
  Or use debugging tools: https://resthooks.io/docs/guides/debugging
  Learn more about schemas: https://resthooks.io/docs/api/schema

  Entity: ${this.name}
  Value: ${input && JSON.stringify(input, null, 2)}
  `,
        );
        (error as any).status = 400;
        throw error;
      } else {
        // these make the keys get deleted
        return undefined;
      }
    }
    const entityType = this.key;

    if (!(entityType in visitedEntities)) {
      visitedEntities[entityType] = {};
    }
    if (!(id in visitedEntities[entityType])) {
      visitedEntities[entityType][id] = [];
    }
    if (
      visitedEntities[entityType][id].some((entity: any) => entity === input)
    ) {
      return id;
    }
    visitedEntities[entityType][id].push(input);

    Object.keys(this.schema).forEach(key => {
      if (Object.hasOwnProperty.call(processedEntity, key)) {
        const schema = this.schema[key];
        processedEntity[key] = visit(
          processedEntity[key],
          processedEntity,
          key,
          schema,
          addEntity,
          visitedEntities,
        );
      } else if (process.env.NODE_ENV !== 'production') {
        const error = new Error(
          `Schema key is missing in Entity

  Be sure all schema members are also part of the entity
  Or use debugging tools: https://resthooks.io/docs/guides/debugging
  Learn more about nesting schemas: https://resthooks.io/docs/guides/nested-response

  Entity keys: ${Object.keys(processedEntity)}
  Schema key(missing): ${key}
  `,
        );
        (error as any).status = 400;
        throw error;
      }
    });

    addEntity(this, processedEntity, id);
    return id;
  }

  static expiresAt(
    { expiresAt }: { expiresAt: number; date: number },
    input: any,
  ): number {
    return expiresAt;
  }

  /** Return true to merge incoming data; false keeps existing entity */
  static useIncoming(
    existingMeta: { date: number; fetchedAt?: number },
    incomingMeta: { date: number; fetchedAt?: number },
    existing: any,
    incoming: any,
  ) {
    return existingMeta.date <= incomingMeta.date;
  }

  static infer(
    args: readonly any[],
    indexes: Record<string, any>,
    infer: any,
  ): any {
    if (!args[0]) return undefined as any;
    const id = this.pk(args[0], undefined, '');
    // Was able to infer the entity's primary key from params
    if (id !== undefined && id !== '') return id as any;
    // now attempt lookup in indexes
    const indexName = indexFromParams(args[0], this.indexes);
    if (indexName && indexes[this.key]) {
      // 'as Record<string, any>': indexName can only be found if params is a string key'd object
      return indexes[this.key][indexName][
        (args[0] as Record<string, any>)[indexName]
      ] as any;
    }
    return undefined as any;
  }

  static denormalize<T extends typeof SimpleRecord>(
    this: T,
    input: Readonly<Partial<AbstractInstanceType<T>>> & { [k: string]: any },
    unvisit: schema.UnvisitFunction,
  ): [AbstractInstanceType<T>, boolean, boolean] {
    // TODO: remove immutable case once we stop storing instances in normalized cache
    const entityCopy: any = isImmutable(input)
      ? (input as any)
      : this.fromJS(
          input instanceof SimpleRecord
            ? this.toObjectDefined(input as any)
            : input,
        );
    // Need to set this first so that if it is referenced further within the
    // denormalization the reference will already exist.
    unvisit.setLocal?.(entityCopy);

    // TODO: this entire function is redundant with SimpleRecord, however right now we're storing the Entity instance
    // itself in cache. Once we offer full memoization, we will store raw objects and this can be consolidated with SimpleRecord
    if (isImmutable(input)) {
      const [denormEntity, found, deleted] = denormalizeImmutable(
        this.schema,
        input,
        unvisit,
      );
      return [this.fromJS(denormEntity.toObject()), found, deleted];
    }
    let deleted = false;

    // note: iteration order must be stable
    Object.keys(this.schema).forEach(key => {
      const schema = this.schema[key];
      const nextInput = Object.hasOwnProperty.call(input, key)
        ? input[key]
        : undefined;
      const [value, , deletedItem] = unvisit(nextInput, schema);

      if (
        deletedItem &&
        !(Object.hasOwnProperty.call(input, key) && !this.defaults[key])
      ) {
        deleted = true;
      }
      if (Object.hasOwnProperty.call(input, key) && input[key] !== value) {
        entityCopy[key] = value;
      }
    });

    return [entityCopy, true, deleted];
  }
}

if (process.env.NODE_ENV !== 'production') {
  // for those not using TypeScript this is a good catch to ensure they are defining
  // the abstract members
  Entity.fromJS = function fromJS<T extends typeof SimpleRecord>(
    this: T,
    props: Partial<AbstractInstanceType<T>>,
  ): AbstractInstanceType<T> {
    if ((this as any).prototype.pk === undefined)
      throw new Error('cannot construct on abstract types');
    return SimpleRecord.fromJS.call(this, props) as any;
  };
}

export function isEntity(schema: Schema): schema is typeof Entity {
  return schema !== null && (schema as any).pk !== undefined;
}

function indexFromParams<I extends string>(
  params: Readonly<object>,
  indexes?: Readonly<I[]>,
) {
  if (!indexes) return undefined;
  return indexes.find(index => Object.hasOwn(params, index));
}
