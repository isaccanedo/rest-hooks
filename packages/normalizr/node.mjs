export * from './dist/normalizr.js';
