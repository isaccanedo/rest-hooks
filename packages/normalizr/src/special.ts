export const DELETED = Symbol('ENTITY WAS DELETED');
