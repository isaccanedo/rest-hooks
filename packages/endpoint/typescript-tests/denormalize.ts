import { normalize, denormalize } from '@rest-hooks/normalizr';
import { IDEntity } from '__tests__/new';

import { schema, DenormalizeNullable, Normalize, Denormalize } from '../src';

class Magic extends IDEntity {
  readonly a = 'first' as const;
  private b = false;
  c = 'arg' as const;
}
class Magic2 extends IDEntity {
  readonly a = 'second' as const;
  private b3 = false;
}
const magicSchema = Magic;
const magic2Schema = Magic2;

const unionSchema = new schema.Union(
  {
    user: magicSchema,
    group: magic2Schema,
  },
  'a',
);
const errorUnionSchema = new schema.Union(
  {
    user: magicSchema,
    group: magic2Schema,
  },
  // @ts-expect-error
  'notexistant',
);
const scheme = {
  thing: { data: unionSchema, members: [Magic] },
  first: '',
  second: '',
};
const schemeEntity = Magic;

const de = denormalize({}, scheme, {});
const r = normalize({}, scheme);

type A = DenormalizeNullable<typeof scheme>;
type B = A['thing']['members'];
type C = DenormalizeNullable<typeof schemeEntity>;
type D = ReturnType<typeof unionSchema['_denormalizeNullable']>;
type F = Denormalize<typeof unionSchema>;
type E = Normalize<typeof scheme>['thing']['data'];

if (de[1]) {
  const value = de[0];
  const piece = value.thing.data?.a;
  const first: string = value.first;
  const members = value.thing.members;
} else {
  const value = de[0];
  const data = value.thing.data;
  const members = value.thing.members;
}
const members2 = de[0].thing.members;

const schemeValues = new schema.Values({ btc: Magic, eth: Magic2 });
const schemeValuesSimple = new schema.Values(Magic);
const [valueValues, foundValues] = denormalize({}, schemeValues, {});
Object.keys(schemeValues).forEach(k => {
  const v = valueValues[k];
  if (v?.a === 'second') {
    const b: Magic2 = v;
  }
});

const [valueValuesSimple, foundValuesSimple] = denormalize(
  {},
  schemeValuesSimple,
  {},
);
