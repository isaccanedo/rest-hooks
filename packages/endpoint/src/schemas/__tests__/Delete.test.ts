// eslint-env jest
import { fromJS } from 'immutable';
import { normalize } from '@rest-hooks/normalizr';
import { IDEntity } from '__tests__/new';

import denormalize from './denormalize';
import { schema } from '../../';
import Entity from '../Entity';
import { DELETED } from '../../special';

let dateSpy: jest.SpyInstance;
beforeAll(() => {
  dateSpy = jest
    // eslint-disable-next-line no-undef
    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});

describe(`${schema.Delete.name} normalization`, () => {
  test('throws if not given an entity', () => {
    // @ts-expect-error
    expect(() => new schema.Delete()).toThrow();
  });

  test('normalizes an object', () => {
    class User extends IDEntity {}

    expect(
      normalize({ id: '1', type: 'users' }, new schema.Delete(User)),
    ).toMatchSnapshot();
  });

  test('normalizes already processed entities', () => {
    class MyEntity extends IDEntity {}
    expect(normalize('1', new schema.Delete(MyEntity))).toMatchSnapshot();
  });

  test('does not infer', () => {
    class User extends IDEntity {}

    expect(
      new schema.Delete(User).infer([{ id: 5 }], {}, () => undefined),
    ).toBeUndefined();
  });

  it('should throw a custom error if data does not include pk', () => {
    class MyEntity extends Entity {
      readonly name: string = '';
      readonly secondthing: string = '';
      pk() {
        return this.name;
      }
    }
    function normalizeBad() {
      normalize({ secondthing: 'hi' }, new schema.Delete(MyEntity));
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });
});

describe(`${schema.Delete.name} denormalization`, () => {
  class User extends IDEntity {
    readonly username: string = '';
    readonly type: string = '';
  }
  const entities = {
    User: {
      '1': User.fromJS({ id: '1', username: 'Janey', type: 'users' }),
    },
  };

  test('denormalizes an object in the same manner as the Entity', () => {
    const [user, ready, deleted] = denormalize(
      '1',
      new schema.Delete(User),
      entities,
    );
    expect(user).toBeDefined();
    expect(user).toBeInstanceOf(User);
    expect(user?.username).toBe('Janey');
    expect(ready).toBe(true);
    expect(deleted).toBe(false);
  });

  test.each([
    ['direct', (data: any) => data],
    ['immutable', fromJS],
  ] as const)(
    `denormalizes deleted entities as undefined (%s)`,
    (_, createInput) => {
      const [user, ready, deleted] = denormalize('1', new schema.Delete(User), {
        User: { '1': DELETED },
      });
      expect(user).toBe(undefined);
      expect(ready).toBe(true);
      expect(deleted).toBe(true);

      expect(
        denormalize(
          createInput([{ data: '1' }]),
          new schema.Array(
            new schema.Object({ data: new schema.Delete(User) }),
          ),
          createInput([{ User: { '1': DELETED } }]),
        ),
      ).toMatchSnapshot();

      expect(
        denormalize(
          createInput([{ data: '1' }]),
          [{ data: new schema.Delete(User) }],
          createInput([{ User: { '1': DELETED } }]),
        ),
      ).toMatchSnapshot();

      expect(
        denormalize(
          createInput([{ data: '1' }]),
          new schema.Array(new schema.Object({ data: User })),
          createInput([{ User: { '1': DELETED } }]),
        ),
      ).toMatchSnapshot();

      expect(
        denormalize(
          createInput([{ data: '1' }]),
          [{ data: User }],
          createInput([{ User: { '1': DELETED } }]),
        ),
      ).toMatchSnapshot();

      expect(
        denormalize(
          createInput({ data: '1' }),
          new schema.Object({ data: User }),
          createInput({ User: { '1': DELETED } }),
        ),
      ).toMatchSnapshot();

      expect(
        denormalize(
          createInput({ data: '1' }),
          { data: User },
          createInput({ User: { '1': DELETED } }),
        ),
      ).toMatchSnapshot();
    },
  );
});
