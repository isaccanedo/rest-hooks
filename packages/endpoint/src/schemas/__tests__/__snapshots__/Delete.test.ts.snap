// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delete denormalization denormalizes deleted entities as undefined (direct) 1`] = `
[
  [
    {
      "data": undefined,
    },
  ],
  true,
  false,
]
`;

exports[`Delete denormalization denormalizes deleted entities as undefined (direct) 2`] = `
[
  [
    {
      "data": undefined,
    },
  ],
  true,
  false,
]
`;

exports[`Delete denormalization denormalizes deleted entities as undefined (direct) 3`] = `
[
  [
    {
      "data": undefined,
    },
  ],
  true,
  false,
]
`;

exports[`Delete denormalization denormalizes deleted entities as undefined (direct) 4`] = `
[
  [
    {
      "data": undefined,
    },
  ],
  true,
  false,
]
`;

exports[`Delete denormalization denormalizes deleted entities as undefined (direct) 5`] = `
[
  {
    "data": undefined,
  },
  true,
  true,
]
`;

exports[`Delete denormalization denormalizes deleted entities as undefined (direct) 6`] = `
[
  {
    "data": undefined,
  },
  true,
  true,
]
`;

exports[`Delete denormalization denormalizes deleted entities as undefined (immutable) 1`] = `
[
  Immutable.List [
    Immutable.Map {
      "data": undefined,
    },
  ],
  true,
  false,
]
`;

exports[`Delete denormalization denormalizes deleted entities as undefined (immutable) 2`] = `
[
  Immutable.List [
    Immutable.Map {
      "data": undefined,
    },
  ],
  true,
  false,
]
`;

exports[`Delete denormalization denormalizes deleted entities as undefined (immutable) 3`] = `
[
  Immutable.List [
    Immutable.Map {
      "data": undefined,
    },
  ],
  true,
  false,
]
`;

exports[`Delete denormalization denormalizes deleted entities as undefined (immutable) 4`] = `
[
  Immutable.List [
    Immutable.Map {
      "data": undefined,
    },
  ],
  true,
  false,
]
`;

exports[`Delete denormalization denormalizes deleted entities as undefined (immutable) 5`] = `
[
  Immutable.Map {
    "data": undefined,
  },
  true,
  true,
]
`;

exports[`Delete denormalization denormalizes deleted entities as undefined (immutable) 6`] = `
[
  Immutable.Map {
    "data": undefined,
  },
  true,
  true,
]
`;

exports[`Delete normalization normalizes already processed entities 1`] = `
{
  "entities": {},
  "entityMeta": {},
  "indexes": {},
  "result": "1",
}
`;

exports[`Delete normalization normalizes an object 1`] = `
{
  "entities": {
    "User": {
      "1": Symbol(ENTITY WAS DELETED),
    },
  },
  "entityMeta": {
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": "1",
}
`;

exports[`Delete normalization should throw a custom error if data does not include pk 1`] = `
"Missing usable primary key when normalizing response.

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.
  Or use debugging tools: https://resthooks.io/docs/guides/debugging
  Learn more about schemas: https://resthooks.io/docs/api/schema

  Delete(Entity): Delete(MyEntity)
  Value: {
  "secondthing": "hi"
}
  "
`;
